File: Service/src/main/java/com/github/spb/tget/demo/ClientManager.java
package com.github.spb.tget.demo;

import com.github.spb.tget.demo.converter.ClientConverter;
import com.github.spb.tget.demo.dao.ClientDao;
import com.github.spb.tget.demo.dao.ContactDao;
import com.github.spb.tget.demo.data.ClientEntity;
import com.github.spb.tget.demo.dto.ClientDto;
import com.github.spb.tget.demo.dto.ContactDto;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.stream.Collectors;

public class ClientManager {

    private ClientDao clientDao = new ClientDao();
    private ContactDao contactDao = new ContactDao();
    private ClientConverter clientConverter = new ClientConverter();

    public List<ClientDto> getClients() {
        return clientDao.getClients().stream()
                .map(client -> clientConverter.toDto(client))
                .collect(Collectors.toList());
    }

    public ClientDto getClient(Integer id) {
        ClientEntity targetClient = clientDao.resolveClient(id);
        if (targetClient == null) {
            throw new NoSuchElementException("Failed to find client with id: " + id);
        }
        return clientConverter.toDto(targetClient);
    }

    public List<ContactDto> getClientContacts(Integer id) {
        return clientConverter.toDto(clientDao.resolveClient(id)).getContacts();
    }

    public ClientDto createClient(ClientDto client) {
        return clientConverter.toDto(clientDao.createClient(clientConverter.fromDto(client)));
    }

    public void addContacts(List<ContactDto> contacts, Integer id) {
        ClientEntity client = clientDao.resolveClient(id);
        contactDao.addContacts(contacts.stream().map(c -> clientConverter.contactInfoFromDto(c))
                .collect(Collectors.toList()), client);
    }

    public void deleteClient(Integer id) {
        ClientEntity client = clientDao.resolveClient(id);
        if (client == null) {
            throw new NoSuchElementException("Failed to find client with id: " + id);
        }
        clientDao.deleteClient(client);
    }

    public void updateClient(Integer id, ClientDto clientDto) {
        ClientEntity client = clientDao.resolveClient(id);
        if (client == null) {
            throw new NoSuchElementException("Failed to find client with id: " + id);
        }
        clientDto.setId(id);
        clientDao.updateClient(clientConverter.fromDto(clientDto));
    }
}

File: Service/src/main/java/com/github/spb/tget/demo/ClientManager.kt
package com.github.spb.tget.demo

import com.github.spb.tget.demo.dao.ClientDao
import com.github.spb.tget.demo.dao.ContactDao
import com.github.spb.tget.demo.data.AlreadyExistsException
import com.github.spb.tget.demo.data.ClientEntity
import com.github.spb.tget.demo.data.ContactEntity
import com.github.spb.tget.demo.data.EntityNotFoundException
import com.github.spb.tget.demo.data.toDto
import com.github.spb.tget.demo.dto.ClientDto
import com.github.spb.tget.demo.dto.ContactDto

class ClientManager {

    private val clientDao = ClientDao()
    private val contactDao = ContactDao()

    fun getClients(): List<ClientDto> {
        return clientDao.getClients().map { it -> it.toDto() }
    }

    fun getClient(id: Int): ClientDto {
        val targetClient = clientDao.resolveClient(id) ?: throw EntityNotFoundException("Client ID $id")
        return targetClient.toDto()
    }

    fun getClientContacts(id: Int): List<ContactDto>? {
        val client = clientDao.resolveClient(id)
        client ?: throw EntityNotFoundException("Client with ID $id")
        return clientDao.resolveClient(id)?.toDto()?.contacts
    }

    fun createClient(client: ClientDto): ClientDto {
        if (clientDao.getClients().firstOrNull { it.clientId == client.id } != null) {
            throw AlreadyExistsException("Client with Id ${client.id} already exists")
        }
        return clientDao.createClient(ClientEntity.fromDto(client)).toDto()
    }

    fun addContacts(contacts: List<ContactDto>, id: Int) {
        val client = clientDao.resolveClient(id) ?: throw EntityNotFoundException("Client ID $id")
        if (contacts.any { contact ->
            contactDao.getAllContacts().any { it.email?.equals(contact.emailAddress, ignoreCase = true) == true }
        }) {
            throw AlreadyExistsException("")
        }

        contactDao.addContacts(contacts.map { it -> ContactEntity.fromDto(it) }, client)
        client.contactInformation?.addAll(contacts.map { it -> ContactEntity.fromDto(it) });
        clientDao.updateClient(client)
    }

    fun deleteClient(id: Int) {
        val client = clientDao.resolveClient(id) ?: throw EntityNotFoundException("Client ID $id")
        clientDao.deleteClient(client)
    }

    fun updateClient(id: Int, clientDto: ClientDto) {
        clientDao.resolveClient(id) ?: throw EntityNotFoundException("Client ID $id")
        clientDto.id = id
        clientDao.updateClient(ClientEntity.fromDto(clientDto))
    }
}